### ISSUE TITLE
BIO_sock_init() needs a concurrency lock on Win32.

### ISSUE THREAD

**Issue by msdsgn** [openssl/openssl#7346]
Opening BIO connections with **BIO_do_connect()** the first time will cause **BIO_sock_init()** to fire up the OS sockets library. On Windows, that is obviously **WSAStartup()**. When you fire up multiple uses of  **BIO_do_connect()** on different threads simulataniously. The first call in one thread will wait for initialisation and succeed, whilst the other connection attempts in other threads will fail with **ERR_R_SYS_LIB** and **BIO_F_BIO_LOOKUP**, probably because the OS is still initialising the library. This problem can be easily resolved by manually calling **BIO_sock_init()** before firing up any threads and calling **BIO_do_connect()**, but perhaps the devs here could just put a simple concurrency lock in **BIO_sock_init()** so multiple requests of this function will wait until the initial request has completed and just silently succeed after.

_Using Windows 8.1 (64-bit) and OpenSSL 1.1.0i (32/64-bit) on MSVisualC 19.13.26128._

**Comment by rbrus** [comment]
I was actually wondering if usage of BIO_do_connect() makes any sense. If it might be better to stick with BIO_connect() , BIO_socket(), etc.

**Comment by bplu4t2f** [comment]
This is still an issue.

Another way to repro is starting two separate threads on program startup, both of which do `BIO_accept` on an accept bio.

One of them will almost certainly fail with `Either the application has not called WSAStartup, or WSAStartup failed.`.

Adding `BIO_sock_init` before starting the thread fixes the issue. This should at least be documented or otherwise mentioned as a *really important* initialization call on Windows (e.g. in the wiki).

**Comment by t8m** [comment]
> This is still an issue.

I'd suggest opening a new issue as the reporter has closed this.
